[MVC - Step 2] 효오 미션 제출합니다.

hyojaekim opened the pull request:
안녕하세요. 효오입니다.

pr 보내고 계속해서 테스트 추가하고, ViewResolver, ArgumentResolver 를 구현할 예정입니다.

항상 좋은 리뷰 주셔서 감사합니다! :)

hyojaekim committed:
feat: JsonView 구현

hyojaekim committed:
feat: 어노테이션 기반 회원가입, 조회, 수정하는 기능 추가

- UserAcceptanceTest 테스트 성공

hyojaekim committed:
refactor: getBody 메서드 로직 변경

hyojaekim committed:
feat: 기존 컨트롤러 -> 어노테이션 기반 HomeController 추가

hyojaekim committed:
feat: 기존 컨트롤러 -> 어노테이션 기반 UserController 추가

- user list 페이지 이동
- user profile 페이지 이동
- user 수정 페이지 이동
- user 정보 수정 기능 추가

hyojaekim committed:
refactor: 기존 레거시 컨트롤러 제거

hyojaekim committed:
feat: 기존 컨트롤러 로직 -> 어노테이션 기반으로 변경

hyojaekim committed:
test: CreateUserControllerTest 추가

- 회원가입 페이지 이동
- 회원가입 시 유저 생성

hyojaekim committed:
test: LoginControllerTest 추가

- 로그인 페이지 이동하는 테스트
- 로그인 성공 테스트
- 로그인 실패 테스트

jeonghoon1107 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/JsonView.java:
```diff
@@ -1,12 +1,43 @@
+        response.getWriter().write(json);
+    }
+
+    private String toJson(Map<String, ?> model) throws JsonProcessingException {
```
JsonUtils라는 클래스가 있는데 해당 클래스로 옮겨보면 어떨까요 :)

hyojaekim replied:
놓친 부분이 있었네요! 피드백 감사합니다.

jeonghoon1107 commented on slipp/src/main/java/slipp/controller/HomeController.java:
```diff
@@ -1,15 +1,24 @@
+
+    @RequestMapping(value = "/", method = RequestMethod.GET)
+    public ModelAndView home(HttpServletRequest request, HttpServletResponse response) {
+        request.setAttribute("users", DataBase.findAll());
```
layer 를 분리해보면 어떨까요?
https://wckhg89.tistory.com/13 (layer 부분만 참고하셔서 보면 좋을 것 같아요!)

hyojaekim replied:
Service Layer 만드려고 생각은 했으나, MVC 프레임워크에 집중하려고 했습니다!
추가 미션으로 ArgumentResolver를 구현하고, Service Layer는 일부만 구현했습니다 :)

jeonghoon1107 reviewed (changes requested):
안녕하세요 효오!
코드가 깔끔하네요 💯 
몇가지 피드백 남겼으니 확인부탁드려요 :)

hyojaekim committed:
feat: HandlerAdapter 추가

- add ControllerHandlerAdapter
- add HandlerExecutionHandlerAdapter

hyojaekim committed:
feat: HandlerAdapter 적용

- DispatcherServlet에 HandlerAdapter 로직 적용
- Adapter를 찾지 못할 경우 예외 추가
- HandlerMapping 리턴 타입 변경

hyojaekim committed:
refactor: (toJson 메소드) JsonView -> JsonUtils 이동

hyojaekim committed:
feat: CreateUserService 추가

hyojaekim committed:
refactor: CreateUserService 리턴 타입 변경

- String -> UserCreateDto

hyojaekim committed:
style: Controller, Service 클래스 이름 수정

- UserCreateController -> CreateUserController
- CreateUserService -> UserCreateService

hyojaekim committed:
test: UserCreateService 유저 생성 테스트 추가

hyojaekim committed:
feat: LoginService, UserSearchService 추가

- 로그인 체크 기능 추가
- 유저 찾는 기능 추가

hyojaekim committed:
test: LoginService 테스트 추가

- 로그인 성공
- 로그인 실패

hyojaekim committed:
fix: 로그인 세션 버그 수정

hyojaekim committed:
feat: ArgumentResolver 적용

hyojaekim committed:
test: ArgumentResolverTest 추가

hyojaekim commented:
안녕하세요. 정훈님!
이번에 MVC 프레임워크를 구현하면서 요구사항 외에 다른 것들을 구현하려고 노력해봤습니다.
ArgumentResolver를 만들었는데 이게 맞는건지 잘 모르겠네요..
시간이 걸릴까봐 일단은 Service Layer는 일부만 만들고 제출하고, 계속해서 리팩토링 하겠습니다.

jeonghoon1107 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/ArgumentResolver.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    private Object matchParameter(Parameter parameter) {
+        if (parameter.isAnnotationPresent(RequestParam.class)) {
```
ArgumentResolver 역시 handlerAdapters와 같이 interface로 구현하면 추가될 때마다 분기를 없앨 수 있을 것 같아요!

hyojaekim replied:
HandlerAdapter 처럼 interface로 구현을 했는데,
ArgumnetResolver의 Adapter들이 네이밍이 적절한지 궁금합니다!

jeonghoon1107 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/ArgumentResolver.java:
```diff
@@ -0,0 +1,65 @@
+        this.response = response;
+    }
+
+    public Object[] resolve(Method method) {
```
parameter type이 맞지 않으면 4xx 예외를 주는 것은 어떨까요?

hyojaekim replied:
**클라이언트에서 유효한 요청**을 받은 다음
해당 메소드를 찾고 **메소드의 parameter를 가져오는 역할**을 하는데
피드백 내용을 토대로 **일치하지 않는 parameter에 대해서 예외**를 던지도록 수정했습니다.

그런데 **클라이언트에서 유효한 요청**을 보냈기 때문에 **4xx 클라이언트 요청에 대한 오류보다**는 
**서버 오류**가 맞겠다고 싶어서 500 에러로 처리하려고 합니다. 괜찮나요?!

```java
private Object[] getParameters(HttpServletRequest request, HttpServletResponse response) {
        try {
            ArgumentResolver argumentResolver = new ArgumentResolver(request, response);
            return argumentResolver.resolve(method);
        } catch (NotMatchParameterException e) {
            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
            throw new NotMatchParameterException();
        }
    }
```

또 상태 코드를 설정할 때, 위와 같이 상태 코드를 설정하도록 변경 했는데 어떤지 궁금합니다.

jeonghoon1107 replied:
스프링에서 RequestParam 어노테이션을 사용할 때 500 서버에러를 상태코드로 받나요?
서버에러는 말 그대로 서버에서 예기치 못한 에러가 발생했을 때 나오는 것이죠!

http status code는 client 측에 우리가 약속된 http 상태 코드로 어떤 부분이 잘못되었다는 것인지 알려주기 위해서 사용하죠!

`@RequestParam("userId") String userId`
url?userI=abc 로 요청을 보냈을 경우 500에러를 반환해주어야 할까요?

jeonghoon1107 commented on nextstep-mvc/src/test/java/nextstep/mvc/tobe/ArgumentResolverTest.java:
```diff
@@ -0,0 +1,49 @@
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ArgumentResolverTest {
```
조금 더 다양한 케이스가 있으면 좋을 것 같아요!

jeonghoon1107 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/handlerAdapter/HandlerExecutionHandlerAdapter.java:
```diff
@@ -0,0 +1,25 @@
+            return new ModelAndView(result.toString());
+        }
+
+        return (ModelAndView) result;
```
type casting하지 않아도 handle의 return이 ModelAndView인 것 같은데 맞나요??

hyojaekim replied:
같이 수정을 했어야 했는데 피드백 감사합니다!

jeonghoon1107 reviewed (changes requested):
안녕하세요 효오!
깔끔하네요 👍 
몇가지 피드백 남겼으니 확인부탁드려요 :)

hyojaekim committed:
refactor: HandlerExecutionHandlerAdapter 불필요한 로직 제거

hyojaekim committed:
feat: ArgumentResolverAdapter 추가

hyojaekim committed:
feat: RequestResolverAdapter 구현

hyojaekim committed:
feat: ResponseResolverAdapter 구현

hyojaekim committed:
feat: RequestParamResolverAdapter 구현

hyojaekim committed:
feat: ArgumnetResolverAdapter 적용

hyojaekim committed:
refactor: 메소드 분리 및 예외 추가

- NotMatchParameterException

hyojaekim committed:
test: ArgumentResolver 테스트 추가

- ArgumentResolver 생성 테스트
- 파라미터 타입 안맞을 때, 예외 발생하는 테스트

hyojaekim committed:
style: ArgumentResolver 패키지 생성 및 이동

hyojaekim committed:
refactor: UserCreateService addUser 리턴 타입 변경

- UserCreatedDto -> String
- 변경된 리턴 타입에 맞게 테스트 수정

hyojaekim committed:
fix: 실패하는 테스트 수정

jeonghoon1107 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/argumentResolver/RequestResolverAdapter.java:
```diff
@@ -0,0 +1,24 @@
+    public boolean match(Parameter parameter) {
+        Class<?> parameterType = parameter.getType();
+
+        return parameterType.equals(HttpServletRequest.class);
```
isAssignableFrom를 사용해보면 어떨까요?
`HttpServletRequest.class.isAssignableFrom(parameterType)`
NPE가 발생할 수 있는 곳에는 항상 유의하면 좋을 것 같아요!

jeonghoon1107 commented on nextstep-mvc/src/main/java/nextstep/mvc/tobe/argumentResolver/ArgumentResolver.java:
```diff
@@ -0,0 +1,42 @@
+import java.lang.reflect.Parameter;
+import java.util.*;
+
+public class ArgumentResolver {
```
객체를 argument로 받을 수 있도록 처리하면 좋았을텐데 아쉽네요!

jeonghoon1107 reviewed (approved):
안녕하세요 효오!
인자 매핑까지 깔끔하게 구현해주셨네요 💯 
관련하여 몇가지 간단한 피드백과 질문해주신 것에 대한 답변 남겼습니다!
미션 진행하느라 고생많으셨습니다 :)
추가적인 질문은 언제든 DM 주세요!
